src/api/app/models/workflow.rb

```diff
@@ -68,4 +81,32 @@ def initialize_step(step_name, step_instructions)
   def supported_filters
     @supported_filters ||= workflow_instructions[:filters]&.select { |key, _value| SUPPORTED_FILTERS.include?(key.to_sym) }
   end
+
+  def destroy_target_projects
+    # Do not process steps for which there's nothing to do
+    processable_steps = steps.reject { |step| step.instance_of?(::Workflow::Step::ConfigureRepositories) }
+    target_packages = steps.map(&:target_package).uniq.compact
+    EventSubscription.where(channel: 'scm', token: self, package: target_packages).delete_all
+
+    target_project_names = processable_steps.map(&:target_project_name).uniq.compact
+    # We want the callbacks to run after destroy, so we can't use delete_all
+    Project.where(name: target_project_names).destroy_all
+  end
+
+  def restore_target_projects(user)
+    token_user_login = user.login
```

Comment:
This goes along with my previous comment in `Token::Workflow`, we should use the token user:

```suggestion
    token_user_login = token.user.login
```